lib/debug/thread_client.rb

```diff
@@ -348,10 +348,9 @@ def instance_eval_for_cmethod frame_self, src
     def frame_eval src, re_raise: false
       @success_last_eval = false
 
-      b = current_frame&.eval_binding || TOPLEVEL_BINDING
-      b = b.dup
+      b = current_frame&.binding || TOPLEVEL_BINDING
 
-      special_local_variables current_frame do |name, var|
+      special_local_variables current_frame, binding: b do |name, var|
```

Comment:
To evaluate in C frames, we should be able to fallback to `TOPLEVEL_BINDING`. So relying on the frame's binding is not enough here.